src/libraries/System.Formats.Cbor/src/System/Formats/Cbor/Writer/CborWriter.Simple.cs

```diff
@@ -130,7 +142,7 @@ public void WriteSimpleValue(CborSimpleValue value)
         private static bool TryConvertDoubleToSingle(double value, out float result)
         {
             result = (float)value;
-            return BitConverter.DoubleToInt64Bits(result) == BitConverter.DoubleToInt64Bits(value);
+            return double.IsNaN(value) || BitConverter.DoubleToInt64Bits(result) == BitConverter.DoubleToInt64Bits(value);
```

Comment:
Is this simply trying to see if a given `double` can be represented as a `float` without loss in precision?

Is there a non-canonicalization mode where the `NaN` payload bits are necessary to consider?

Reply:
> Is this simply trying to see if a given `double` can be represented as a `float` without loss in precision?

That's what the hitherto bit pattern comparison was trying to do. That failed for FPUs with lax NaN propagation rules (e.g. RISC-V) because the conversion to float drops the sign and payload.

> Is there a non-canonicalization mode where the `NaN` payload bits are necessary to consider?

[RFC 8949](https://www.rfc-editor.org/rfc/rfc8949) offers an easy way out by saying "we don't support payloads":
*If NaN is an allowed value, and there is no intent to support NaN payloads or signaling NaNs, the protocol needs to pick a single representation, typically 0xf97e00. If that simple choice is not possible, specific attention will be needed for NaN handling.*

However, CborConformanceMode.Ctap2CanonicalMode "that simple choice is not possible" as it says: *representations of any floating-point values are not changed*. For that mode we don't enter TryConvertDoubleToSingle and friends so no problem there. But I guess in Write(Single|Double)Core and WriteHalf the condition should be:
```cs
if (float.IsNaN(value) && !CborConformanceModeHelpers.RequiresPreservingFloatPrecision(ConformanceMode))
{
    value = CanonicalNaNSingle;
}
```
Agreed?